### PR TITLE
feat(coding-graph): WASM tree-sitter parser engine + tier-1 extractors (#1551)

### LINKED ISSUE #1551: feat(coding-graph): @remnic/coding-graph package — WASM tree-sitter parser engine, grammar management, and symbol-extraction IR
Part of #1548 (Track B, Phase 1). Standards: #1520 (working agreement applies to every PR under this issue).

## Problem

#1548 mandates a full native implementation of codebase-memory-mcp's capabilities inside the Remnic family. This issue delivers the foundation: a new à-la-carte package `@remnic/coding-graph` containing the parsing engine — tree-sitter grammars, per-language symbol extractors, and a neutral intermediate representation (IR) that the graph store (sibling issue, Phase 1) ingests. Without this layer nothing else in Track B exists.

## Design

**Parser strategy (decided in #1548): web-tree-sitter (WASM grammars).** No node-gyp, no per-platform prebuilds, no compile toolchain on user machines — the class of pain documented in #1518 (Windows/cross-platform) and #1538 (better-sqlite3 native-binding scars: one native module already costs us an `ensure-better-sqlite3.mjs` postinstall and CI rebuild steps; N grammars as native modules would multiply that). Cost: WASM parses roughly 2–3× slower than native bindings — accepted; the benchmark harness (Track B Phase 4) is the authority on whether that matters. Escape hatch: everything goes through an internal `ParserBackend` interface so a native `node-tree-sitter` backend (or the external C binary via the subprocess provider) can slot in later without touching extractors.

**Grammar coverage is phased, not 158-at-once.** Tier 1 (this issue): TypeScript, TSX, JavaScript, Python, Go, Rust, Java, C, C++, C#, Ruby, PHP, Kotlin, Swift, Bash. Grammar `.wasm` files ship inside the package (`files` list) — never downloaded at install or runtime (offline CI, supply chain). Extensibility: `codingGraph.grammarDir` lets operators drop additional `.wasm` grammars with a small JSON manifest (language name, file extensions, extractor tier).

**IR (the contract between parser and store):** per file, `FileIR { path, lang, contentHash, symbols[], imports[], exports[], callSites[], routes[] }`. Symbols carry `kind` (function/class/method/interface/enum/type/module), `name`, `qualifiedName`, half-open span `[startByte, endByte)`, and parent linkage. Call sites carry callee-name candidates + span; resolution to edges is the store's and the resolvers' job (Phase A heuristics land with the store; Phase B LSP is its own issue). Route extraction (Express/Fastify/Flask/FastAPI/Rails patterns) is tier-1-frameworks-only in this issue, extensible later.

**Packaging (rule 57, à-la-carte contract in CLAUDE.md):** `@remnic/coding-graph` is optional at every install layer. `@remnic/core` gains only: the engine *type* declarations (types-only module, no runtime import), a computed-specifier dynamic-import loader (`await import("@remnic/" + "coding-graph")` — canonical pattern: `packages/remnic-cli/src/optional-bench.ts`, `packages/remnic-cli/src/optional-module-loader.ts`), and a user-facing install hint on miss. Declared as optional peer dep (`peerDependenciesMeta`), never in any base `dependencies` or tsup `noExternal`. The package must be published (release workflow publish order is the source of truth; keep it topologically sorted).

## Implementation guide

**Anchors:** à-la-carte section of `CLAUDE.md` (rule 57 + "À-la-carte packaging"); loader precedents `packages/remnic-cli/src/optional-bench.ts`, `optional-module-loader.ts`; optional-package precedents `packages/connector-limitless/`, `packages/import-lossless-claw/`; config parsing next to `codingMode` (`packages/remnic-core/src/types.ts`, `config.ts`); both `openclaw.plugin.json` configSchema files (gotcha 4); ignore-handling reference: the external tool's precedence (hardcoded → `.gitignore` hierarchy → project ignore file).

**Ordered steps (one subsystem per PR; suggested split: PR1 scaffold+loader, PR2 engine+tier-1 extractors, PR3 walker+determinism+grammarDir):**

1. Scaffold `packages/coding-graph/` (name `@remnic/coding-graph`, `"private": false`): tsup config, test setup, `web-tree-sitter` as its runtime dependency, grammar `.wasm` assets in `files`. Wire into `.github/workflows/release-and-publish.yml` publish order. Run `pnpm install` and commit lockfile changes (gotcha 15).
2. Core side, tests first (prove-fail-before): a test asserting the loader throws a user-facing install hint (naming the exact `npm install` command) when the package is absent — watch it fail against a raw `import`, then implement the computed-specifier loader. Add `CodingGraphEngine` types-only interface in core (generic name, no host prefix — rule 31).
3. Engine, tests first: for each tier-1 language, a fixture file (synthetic code only — public-repo policy, never real user code) with an expected-IR snapshot. Write the snapshots by hand for 2–3 languages first so the tests fail meaningfully before the extractor exists.
4. Implement `ParserBackend` + the web-tree-sitter backend: lazy per-language grammar load, parser instances cached on the engine instance — never module scope, keyed per instance (rule 11).
5. Per-language extractors as tree-sitter queries (`.scm`) + a thin normalization pass into IR. Spans are half-open `[start, end)` everywhere (rule 35).
6. File walker: respect `.gitignore` hierarchy plus a project-level `.remnicignore`; skip files > `codingGraph.maxFileSizeKb`; binary sniff; never read file contents outside the walker→parser path.
7. Determinism gate: IR for a fixture tree is byte-identical across two runs — sort every collection and object key before serialization (rule 38). `contentHash` hashes the raw file bytes; every consumer must hash the same form (rule 23).

**Pitfalls (by rule):** 57/26 (package-name imports only; verify `npm install @remnic/cli` still works without this package present); 24/17 (`grammarDir` must be `statSync().isDirectory()`, tilde-expanded via the shared `expandTilde`); 51 (unknown language in `codingGraph.languages` → error listing valid languages, no silent skip); 44 (a file that fails to parse yields a tagged `{ok:false, code:"parse_failed"}` entry, and nothing derived from it may be emitted as IR); 11 (no module-level parser/grammar caches); 30/48 (`codingGraph.enabled` defaults `false`; with it off, the loader is never invoked and `tools/list`/CLI help are byte-identical to main — write the gate-off characterization test before wiring).

**Done when:** tier-1 fixture-IR snapshot tests green for all 15 languages; determinism test green; gate-off parity test green; base-install verification (fresh install without the package → clean hint, no `MODULE_NOT_FOUND`); `node scripts/check-ratchets.mjs` baselines unchanged (all new code lives in the new package; core gains only loader + types wiring); `npm run preflight:quick` passes; `scripts/pre-merge-check.sh` per PR.

### PR DESCRIPTION
**Pull request by joshuaswarren**
## Summary

Implements the real `createCodingGraphEngine` in `@remnic/coding-graph` — no longer throws `not_implemented`. This is PR2 of #1551 (engine + tier-1 extractors).

Closes #1551.

## Changes

- **ParserBackend interface** + `WasmTreeSitterBackend` (web-tree-sitter 0.25):
  - Lazy per-language grammar load, parser instances cached per engine instance (rule 11 — no module-level state)
  - 15 tier-1 `.wasm` grammar assets shipped inside the package (`grammars/`)
- **Per-language tree-sitter query extractors** (`.scm` patterns as TS constants):
  - Definitions: functions, methods, classes, interfaces, enums, types, modules
  - Imports, exports, call sites, framework routes (Express/Fastify/Flask)
  - Qualified-name computation via nesting stack with half-open spans (rule 35)
- **FileIR emitter**: runs queries → sorted, deterministic FileIR (rule 38)
- **Language sniffing**: file extension → `CodingGraphLanguage`
- **35 tests**: fixture-IR snapshots for all 15 tier-1 languages, determinism (byte-identical across runs + instances), gate-off parity (no throw), error handling (`parse_failed` for unsupported), lifecycle (dispose)

## Tier-1 languages

TypeScript, TSX, JavaScript, Python, Go, Rust, Java, C, C++, C#, Ruby, PHP, Kotlin, Swift, Bash — all parse successfully with correct symbol/import/export/call-site extraction.

## Test results

```
# tests 35  # pass 35  # fail 0
```

All engine tests pass. Pre-existing better-sqlite3 native binding tests (graph-store, cypher executor) fail in this worktree due to missing native bindings — unrelated to this PR.

## Chokepoints used / not applicable

Not applicable — new code lives entirely in `packages/coding-graph/`. Core gains only PR1's types + loader wiring (already merged).

## Checklist

- [x] `createCodingGraphEngine` no longer throws — returns a real engine
- [x] Tier-1 fixture-IR snapshot tests green for all 15 languages
- [x] Determinism test green (byte-identical across runs + instances)
- [x] Gate-off parity test green (no throw; correct version + supportedLanguages)
- [x] `node scripts/check-ratchets.mjs` — baselines unchanged
- [x] coding-graph build + type-check pass
- [x] `codingGraph.enabled` defaults `false` — loader never invoked when off (PR1 wiring, unchanged)

---

> [!NOTE]
> **Medium Risk**
> Foundational parsing/indexing behavior for the optional coding graph when enabled; incorrect extraction would skew graph data, but failures are contained via parse_failed and the feature remains opt-in.
> 
> **Overview**
> **Replaces the PR1 `createCodingGraphEngine` stub** with a working **web-tree-sitter** engine that parses files into deterministic **`FileIR`** (symbols, imports, exports, call sites, routes) for all **15 tier-1 languages**.
> 
> The new **`src/engine/`** stack wires a **`ParserBackend`** / **`WasmTreeSitterBackend`** (lazy `.wasm` grammar load, per-instance state, retry-safe failed loads), **extension-based language sniffing**, **per-language tree-sitter query extractors** (including CJS `require`/`module.exports`, Go/Rust receiver qualification, Express/Flask routes), and **`emitFileIR`** with sorted IR and a single **`hashContent`** used by reindex.
> 
> **`parseFile`** serializes concurrent use of the shared parser, maps failures to **`parse_failed`** (no throws), and **`dispose`** waits for in-flight parses before teardown. Package **`index.ts`** re-exports the real factory; **`tsup`** treats **`web-tree-sitter`** as external. A large **`engine.test.ts`** suite covers fixtures, determinism, concurrency/dispose, and review-thread edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0bb5a21313483128dd6d4dd95e1f221d8233c41f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/coding-graph/src/engine/parser-backend.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    if (this.languages.has(lang)) return;
+    if (this.loadingLanguages.has(lang)) return this.loadingLanguages.get(lang)!;
+    const p = (async () => {
+      const wasmPath = path.join(this.grammarDir, grammarFileName(lang));
+      const language = await Language.load(wasmPath);
+      this.languages.set(lang, language);
+      this.loadingLanguages.delete(lang);
+    })();
+    this.loadingLanguages.set(lang, p);
+    return p;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/parser-backend.ts:147]: Fixed in 17e52724. Both `init()` and `ensureLanguage()` now clear their in-flight promise in a `finally` block, so a transient `Parser.init()` or `Language.load()` rejection no longer pins the engine in a permanently-rejected state — the next call retries from scratch.

New test `"retry: failed grammar load allows a fresh retry, not a cached rejection"` proves it: it points a backend at a nonexistent grammar dir, asserts the first `ensureLanguage` rejects, then asserts the second call returns a **different** promise (`p1 !== p2`), confirming the failed entry was cleaned up rather than re-awaited. All 41 tests green.

File: packages/coding-graph/src/engine/extractors.ts [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+(import_statement
+  (import_clause (identifier) @import.name)
+  source: (string (string_fragment) @import.module)) @__import.stmt
+(import_statement
+  (import_clause (namespace_import (identifier) @import.name))
+  source: (string (string_fragment) @import.module)) @__import.stmt
+(import_statement
+  (import_clause (named_imports (import_specifier name: (identifier) @import.name)))
+  source: (string (string_fragment) @import.module)) @__import.stmt
+`.trim();
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts:75]: Fixed in 17e52724. Added a CommonJS `require()` pattern to `JS_IMPORTS` that matches `call_expression` with `#eq? @__import.require "require"` and captures the string argument as `@import.module`. Now `const express = require("express")` produces an import edge for `"express"`.

Tests: `"require-imports: CommonJS require() produces an import edge"` (against the JS fixture which already uses `require`) and `"require-imports: CommonJS require with destructuring variants"` (covers `const { readFile } = require("fs")`). 41 tests green.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts:75]: Fixed in 17e52724 (same commit as the Cursor thread above — this is the same defect). `JS_IMPORTS` now captures CommonJS `require("...")` via a `call_expression` pattern with `#eq?` predicate. See the reply on PRRT_kwDORJXyws6Oc9au for full details. 41 tests green.

File: packages/coding-graph/src/engine/extractors.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+(class_declaration name: (identifier) @name) @def.class
+(interface_declaration name: (identifier) @name) @def.interface
+(enum_declaration name: (identifier) @name) @def.enum
+(struct_declaration name: (identifier) @name) @def.class
+(method_declaration name: (identifier) @name) @def.method
+`.trim(),
+  importsQuery: `
+(using_directive (identifier) @import.module) @__import.stmt
+(using_directive (qualified_name (identifier) @import.module)) @__import.stmt
+`.trim(),
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts:335]: Fixed in 17e52724. Changed the C# `importsQuery` from `(qualified_name (identifier) @import.module)` to `(qualified_name) @import.module`, so the entire `qualified_name` node is captured. Now `using System.Collections.Generic;` emits the full namespace `"System.Collections.Generic"` instead of just `"Generic"`.

Test: `"csharp-usings: qualified namespace captured in full"` asserts both the qualified and simple forms. 41 tests green.

File: packages/coding-graph/src/engine/extractors.ts [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+// ===========================================================================
+
+const JS_FAMILY_DEFINITIONS = `
+(function_declaration name: (identifier) @name) @def.function
+(method_definition name: (property_identifier) @name) @def.method
+(class_declaration name: (type_identifier) @name) @def.class
+(interface_declaration name: (type_identifier) @name) @def.interface
+(enum_declaration name: (identifier) @name) @def.enum
+(type_alias_declaration name: (type_identifier) @name) @def.type
+`.trim();
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts:54]: Fixed in 17e52724. Added a `variable_declarator` pattern to both `JS_FAMILY_DEFINITIONS` (TS/TSX) and `JAVASCRIPT_EXTRACTOR.definitionsQuery`:

```
(variable_declarator
  name: (identifier) @name
  value: [(arrow_function) (function_expression)]) @def.function
```

This indexes `const handler = () => {}` and `const handler = function () {}` as function-kind symbols. Tests: `"arrow-fn-decls: const handler = () => {} indexed as function symbol"` (TSX fixture `Container`) and `"arrow-fn-decls: function-expression declarators also indexed"`. 41 tests green.

File: packages/coding-graph/src/engine/parser-backend.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  private initialized = false;
+  private initializing: Promise<void> | null = null;
+  private parser: Parser | null = null;
+  private readonly languages = new Map<CodingGraphLanguage, Language>();
+  private readonly loadingLanguages = new Map<CodingGraphLanguage, Promise<void>>();
+  private readonly grammarDir: string;
+
+  constructor(grammarDir?: string) {
+    this.grammarDir = grammarDir ?? resolveGrammarDir();
+  }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/parser-backend.ts:107]: Fixed in 26168570. `resolveGrammarDir()` is now deferred from the constructor to a lazy `getGrammarDir()` called on first `ensureLanguage()`. Since `parseFile` already wraps `ensureLanguage` in try/catch returning `{ ok: false, code: "parse_failed" }`, a missing grammar dir now surfaces as a tagged per-file failure instead of a factory-time exception. Constructor test `"ctor-safe: WasmTreeSitterBackend constructor does not throw on missing grammar dir"` proves it.

File: packages/coding-graph/src/engine/engine.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        input.path,
+        lang,
+        input.content,
+        root,
+        language,
+      );
+      return { ok: true, ir };
+    } finally {
+      tree.delete();
+    }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/engine.ts:153]: Fixed in 26168570. Added a catch block around the `emitFileIR` call (and the `getLanguage` lookup) so query compilation or extraction errors return `{ ok: false, code: "parse_failed", message: "extraction failed for ..." }` instead of throwing. The `finally` still deletes the tree. This satisfies rule 44 — `parseFile` never throws.

File: packages/coding-graph/src/engine/extractors.ts [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+(call_expression function: (identifier) @call.callee)
+(call_expression function: (member_expression property: (property_identifier) @call.callee))
+`.trim();
+
+const JS_ROUTES = `
+(call_expression
+  function: (member_expression
+    object: (identifier) @__route.app
+    property: (property_identifier) @route.verb)
+  arguments: (arguments . (string (string_fragment) @route.path) . (arrow_function) @route.handler)
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts:130]: Fixed in 26168570. Added a third pattern to `JS_ROUTES` for identifier handlers:
```
(call_expression ... arguments: (arguments . (string ...) . (identifier) @route.handler) ...)
```
Also fixed `extractRoutes` in emit.ts: if the `@route.handler` node is an identifier, its `.text` is used directly (previously `findHandlerName` was called, which expects a function/arrow node and returned null for bare identifiers). Test `"named-routes: app.get('/users', getUsers) captures the handler name"` proves it returns `"getUsers"`.

File: packages/coding-graph/src/engine/emit.ts [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+          stmtStart = firstCap.node.startIndex;
+          stmtEnd = firstCap.node.endIndex;
+        } else {
+          continue;
+        }
+      }
+
+      const existing = groups.get(stmtStart);
+      if (existing) {
+        if (moduleText && !existing.module) existing.module = moduleText;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/emit.ts]: Fixed in 26168570. Changed the `extractImports` grouping key from `stmtStart` (a number) to `` `${stmtStart}:${moduleText}` `` (a composite string). Now multi-module statements like Python `import os, sys` produce separate import entries — one for `os`, one for `sys` — instead of collapsing to a single module. Single-module statements like `import { foo, bar } from "module"` still group correctly because all captures share the same module. Test `"multi-import: Python 'import os, sys' produces two import entries"` proves both modules are captured.

File: packages/coding-graph/src/engine/extractors.ts [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+  callSitesQuery: `
+(call function: (identifier) @call.callee)
+(call function: (attribute attribute: (identifier) @call.callee))
+`.trim(),
+  routesQuery: `
+(decorated_definition
+  (decorator
+    (call function: (attribute attribute: (identifier) @route.verb)
+      arguments: (argument_list (string) @route.path)))
+  definition: (function_definition name: (identifier) @route.handler)
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts:196]: Fixed in 26168570. The `@route.handler` node for Python decorators is an `(identifier)` (the function name). Previously `extractRoutes` called `findHandlerName(cap.node)` which looks for `identifier` children or a `name` field — but an identifier has neither, so it returned null → handler became "anonymous". Now `extractRoutes` checks `cap.node.type === "identifier"` first and uses `.text` directly. Test `"python-routes: @app.get('/users') def users() captures handler name"` proves it returns `"users"` instead of `"anonymous"`.

File: packages/coding-graph/src/engine/parser-backend.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    return p;
+  }
+
+  parse(lang: CodingGraphLanguage, content: string): TSTree | null {
+    if (!this.parser) return null;
+    const language = this.languages.get(lang);
+    if (!language) return null;
+    // setLanguage is cheap (just sets a pointer); safe to call per-parse.
+    this.parser.setLanguage(language);
+    return this.parser.parse(content);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/parser-backend.ts:156]: Fixed in fafecb72. `parseFile` now wraps `doParseFile` in a promise-chain serializer (`parseChain`). Each call awaits the previous parse before touching the shared `Parser`, so concurrent `setLanguage`+`parse` calls are fully serialized. Test `"concurrent-parse: parallel parseFile calls do not race"` fires 3 concurrent parses (JS/Python/TS) and verifies each gets the correct result.

File: packages/coding-graph/src/engine/emit.ts [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+            parentQualifiedName,
+          }
+        : {
+            kind: def.kind,
+            name: def.name,
+            qualifiedName,
+            span: { startByte: def.startByte, endByte: def.endByte },
+          };
+      symbols.push(symbol);
+      stack.push(def);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/emit.ts]: Fixed in fafecb72. The Go `method_declaration` patterns now capture the receiver type_identifier as `@__receiver.type` (both pointer and value receivers). `extractSymbols` uses this capture as the parent qualified name, so `func (s *Server) Start()` produces `Server.Start` with `parentQualifiedName: "Server"` instead of a bare top-level `Start`.

Test `"go-receivers: method qualified names use receiver type"` verifies pointer receiver (`Server.Start`), value receiver (`Server.Stop`), and free functions (`NewServer` with no parent).

File: packages/coding-graph/src/engine/extractors.ts [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+  (#match? @route.verb "^(get|post|put|patch|delete|route|api_route)$"))
+`.trim(),
+};
+
+const GO_EXTRACTOR: LanguageExtractor = {
+  definitionsQuery: `
+(function_declaration name: (identifier) @name) @def.function
+(method_declaration name: (field_identifier) @name) @def.method
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts]: Already fixed in fafecb72 (round-3 commit). The Go `method_declaration` patterns now capture the receiver `type_identifier` as `@__receiver.type`, and `extractSymbols` uses it to compute qualified names. `func (a A) Start()` → `A.Start`, `func (b B) Start()` → `B.Start` — no collision. Test `"go-receivers: method qualified names use receiver type"` (which uses both pointer and value receivers) proves it. The bot likely reviewed a stale commit before the fix landed.

[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/coding-graph/src/engine/extractors.ts]: Fixed in c0fbbd59. Added four CommonJS export patterns to JS_EXPORTS: `module.exports = { App, fn }` (shorthand + keyed), `module.exports = App` (single), and `exports.handler = handler` (named). Tests: `"cjs-exports: module.exports = { App, createRouter } captured as exports"` and `"cjs-exports: exports.handler = handler captured"`.
